### PR TITLE
Hash join build projection push down

### PR DESCRIPTION
**Pull request by andyfengHKU**
This PR adds the framework for projection push down optimizer, specifically

- avoids materializing unnecessary properties before hash join build

### CODE REVIEW

File: src/main/connection.cpp [hunk 1]
```diff
@@ -173,8 +173,9 @@ std::unique_ptr<PreparedStatement> Connection::prepareNoLock(
             plans.push_back(Planner::getBestPlan(
                 *database->catalog, nodeStatistics, relStatistics, *boundStatement));
         }
-        for (auto& plan : plans) {
-            optimizer::Optimizer::optimize(plan.get());
+        for (auto i = 0; i < plans.size(); ++i) {
+            auto s = plans[i]->toString();
```
Comment (andyfengHKU) [inline src/main/connection.cpp:177]: revert

File: src/optimizer/projection_push_down_optimizer.cpp [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+void ProjectionPushDownOptimizer::visitHashJoin(planner::LogicalOperator* op) {
+    auto hashJoin = (LogicalHashJoin*)op;
+    for (auto& joinNodeID : hashJoin->getJoinNodeIDs()) {
+        collectPropertiesInUse(joinNodeID);
+    }
+    if (hashJoin->getJoinType() == JoinType::MARK) { // no need to perform push down for mark join.
+        return;
+    }
+    expression_set expressionsBeforePruning;
```
Comment (ray6080) [inline src/optimizer/projection_push_down_optimizer.cpp:70]: It seems you don't need the `expressionsBeforePruning`.
You can loop over hashJoin->getExpressionsToMaterialize() and directly insert into `expressionsAfterPruning` when necessary.

File: src/include/optimizer/projection_push_down_optimizer.h [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    void visitOrderBy(planner::LogicalOperator* op);
+
+    void collectPropertiesInUse(std::shared_ptr<binder::Expression> expression);
+
+private:
+    binder::expression_set propertiesInUse;
+};
+
+} // namespace optimizer
+} // namespace kuzu
```
Comment (ray6080) [inline src/include/optimizer/projection_push_down_optimizer.h:36]: New line here.

**Review by ray6080 — APPROVED** [review]

File: src/optimizer/projection_push_down_optimizer.cpp [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+        std::move(projectionExpressions), hashJoin->getChild(1));
+    hashJoin->setChild(1, std::move(projection));
+}
+
+void ProjectionPushDownOptimizer::visitIntersect(planner::LogicalOperator* op) {
+    auto intersect = (LogicalIntersect*)op;
+    collectPropertiesInUse(intersect->getIntersectNodeID());
+    for (auto i = 0u; i < intersect->getNumBuilds(); ++i) {
+        collectPropertiesInUse(intersect->getBuildInfo(i)->keyNodeID);
+    }
```
Comment (ray6080) [inline src/optimizer/projection_push_down_optimizer.cpp:98]: Add the logic to handle materialized properties in intersect build side.

**Pull request merged**